### PR TITLE
rgw/notification: Make the LifeCycle & Replication events AWS Compatible and Support s3:Replication:Delete event

### PR DESCRIPTION
**Pull request by kchheda3**
Make the LifeCycle and Replication events to be AWS compatible, also at same time support the original event name that was used by ceph for backward compatibility. So if an user subscribes bucket for both lifecycle events (ceph specific and aws compatible), it will get 2 events delivered.
Also as part of changes extend the support for Replication:Delete events.

Fixes tracker --> https://tracker.ceph.com/issues/64251

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [x] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

File: src/rgw/rgw_sal.h [hunk 1]
```diff
@@ -302,9 +302,15 @@ class Driver {
         rgw::notify::EventType event_type, optional_yield y, const std::string* object_name=nullptr) = 0;
```
Comment (yuvalif) [inline src/rgw/rgw_sal.h:302]: why not change this API as well to `rgw::notify::EventTypeList` ?

**Comment by yuvalif** [comment]
as an alternative, you can keep the APIs with a single event type.
the code that encode/decode the event type can normalize to a the AWS version of events.
the downside of this solution would be that if someone configures the ceph event types, they would still see the AWS event type when the look at the event of fetch the configuration.

**Comment by kchheda3** [comment]
> as an alternative, you can keep the APIs with a single event type. the code that encode/decode the event type can normalize to a the AWS version of events. the downside of this solution would be that if someone configures the ceph event types, they would still see the AWS event type when the look at the event of fetch the configuration.

but how will the `notification::match()`, currently each s3 action/LC/replication action sends the event it wants to match, if user sets the aws and we only check for ceph it will not work.
then we also need to hack the `notification::match()` to check additional event if ceph_event is send which IMO is not great (i did think about it, but felt it was not great for readability and future references).
With this workflow you know exactly which events we are checking

**Comment by yuvalif** [comment]
> > as an alternative, you can keep the APIs with a single event type. the code that encode/decode the event type can normalize to a the AWS version of events. the downside of this solution would be that if someone configures the ceph event types, they would still see the AWS event type when the look at the event of fetch the configuration.
> 
> but how will the `notification::match()`, currently each s3 action/LC/replication action sends the event it wants to match, if user sets the aws and we only check for ceph it will not work. then we also need to hack the `notification::match()` to check additional event if ceph_event is send which IMO is not great (i did think about it, but felt it was not great for readability and future references). With this workflow you know exactly which events we are checking

i don't suggest that. i thought we can normalize to the aws-style names - when we decode/encode the event name i nthe conf object.

**Comment by kchheda3** [comment]
> > > as an alternative, you can keep the APIs with a single event type. the code that encode/decode the event type can normalize to a the AWS version of events. the downside of this solution would be that if someone configures the ceph event types, they would still see the AWS event type when the look at the event of fetch the configuration.
> > 
> > 
> > but how will the `notification::match()`, currently each s3 action/LC/replication action sends the event it wants to match, if user sets the aws and we only check for ceph it will not work. then we also need to hack the `notification::match()` to check additional event if ceph_event is send which IMO is not great (i did think about it, but felt it was not great for readability and future references). With this workflow you know exactly which events we are checking
> 
> i don't suggest that. i thought we can normalize to the aws-style names - when we decode/encode the event name i nthe conf object.

So when the user sends `put_bucket_notif` and if user sends the ceph-specific we still store the `aws-compatible` and then during sending of event we just send the aws one ? did i understand correctly?

File: doc/radosgw/s3-notification-compatibility.rst [hunk 2]
```diff
@@ -97,14 +97,30 @@ Event Types
...
+| ``s3:LifecycleExpiration:DeleteMarkerCreated``         | Supported                               |
++--------------------------------------------------------+-----------------------------------------+
+| ``s3:LifecycleTransition``                             | Supported                               |
++--------------------------------------------------------+-----------------------------------------+
 | ``s3:ObjectSynced:*``                                  | Ceph extension                          |
 +--------------------------------------------------------+-----------------------------------------+
 | ``s3:ObjectSynced:Create``                             | Ceph Extension                          |
 +--------------------------------------------------------+-----------------------------------------+
-| ``s3:ObjectSynced:Delete``                             | Defined, Ceph extension (not generated) |
+| ``s3:ObjectSynced:Delete``                             | Ceph extension                          |
```
Comment (jmundack) [inline doc/radosgw/s3-notification-compatibility.rst:112]: should we document the fact that the S3 standard events are the recommended ones? (and long term the duplicate ceph extension notifications will be removed?)

**Comment by yuvalif** [comment]
> > > > as an alternative, you can keep the APIs with a single event type. the code that encode/decode the event type can normalize to a the AWS version of events. the downside of this solution would be that if someone configures the ceph event types, they would still see the AWS event type when the look at the event of fetch the configuration.
> > > 
> > > 
> > > but how will the `notification::match()`, currently each s3 action/LC/replication action sends the event it wants to match, if user sets the aws and we only check for ceph it will not work. then we also need to hack the `notification::match()` to check additional event if ceph_event is send which IMO is not great (i did think about it, but felt it was not great for readability and future references). With this workflow you know exactly which events we are checking
> > 
> > 
> > i don't suggest that. i thought we can normalize to the aws-style names - when we decode/encode the event name i nthe conf object.
> 
> So when the user sends `put_bucket_notif` and if user sends the ceph-specific we still store the `aws-compatible` and then during sending of event we just send the aws one ? did i understand correctly?

yes. thinking about that again, I'm not 100% sure this is the right solution...
it would simplify the fast path (which is alway good), but would require some changes to the test code, and might come as a surprise to users as they will get back a different value than the one they set.

**Review by yuvalif — APPROVED** [review]

**Comment by kchheda3** [comment]
@ceph/doc-writers can you take look at the PR?

File: doc/radosgw/s3-notification-compatibility.rst (hunk 2 above)
Comment (mattbenjamin) [reply doc/radosgw/s3-notification-compatibility.rst:112]: I think deprecating things is jumping the gun

**Review by mattbenjamin — APPROVED** [review]
I have no objections here, but please avoid editorializing in the doc, and proposing there that the initial notification names are being deprecated

**Comment by kchheda3** [comment]
> I have no objections here, but please avoid editorializing in the doc, and proposing there that the initial notification names are being deprecated

we haven't added/updated any doc mentioning about the deprecation

**Comment by yuvalif** [comment]
> > I have no objections here, but please avoid editorializing in the doc, and proposing there that the initial notification names are being deprecated
> 
> we haven't added/updated any doc mentioning about the deprecation

agree we should not deprecate. 
but since we have now 2 sets of event types that do the same thing, we may want to recommend using the more standard ones?

**Comment by yuvalif** [comment]
@kchheda3,
* could you please do a rebase instead of merge, and do the rebase fixes as part of the original commits?
* do we plan on adding a comment in the doc regarding which event names are reccomended (no need to deprecate anything)?

**Comment by mattbenjamin** [comment]
@yuvalif I do not follow what the point is.  There is no need to recommend users to use AWS event names.

**Comment by kchheda3** [comment]
> > @yuvalif I do not follow what the point is. There is no need to recommend users to use AWS event names.
> 
> we have 2 sets of events that do the same thing. if we don't recommend one set, or explain that they are the same, users will ask what is the difference. i think we should add something in the doc to prevent this confusion.

@yuvalif @mattbenjamin  Lemme know if i should go ahead and write the recommended section for sending LC/Replication events in the doc ?

**Comment by yuvalif** [comment]
> > > @yuvalif I do not follow what the point is. There is no need to recommend users to use AWS event names.
> > 
> > 
> > we have 2 sets of events that do the same thing. if we don't recommend one set, or explain that they are the same, users will ask what is the difference. i think we should add something in the doc to prevent this confusion.
> 
> @yuvalif @mattbenjamin Lemme know if i should go ahead and write the recommended section for sending LC/Replication events in the doc ?

the AWS names are more "standard", but, OTOH, they don't cover all of the cases.
so, we should not recommend any of the sets, just say which is equivalent to which in the "note" column. e.g.
```
| s3:LifecycleExpiration:Delete              | Supported. Equivalent to: s3:ObjectLifecycle:Expiration:Current
| s3:LifecycleExpiration:DeleteMarkerCreated | Supported. Equivalent to: s3:ObjectLifecycle:Expiration:DeleteMarker
| s3:LifecycleExpiration:*                   | Supported. Equivalent to: s3:LifecycleExpiration:Delete, s3:LifecycleExpiration:DeleteMarkerCreated
| s3:LifecycleTransition                     | Supported. Equivalent to: s3:ObjectLifecycle:Transition:Current
```

similarly, for the replication event names

**Comment by kchheda3** [comment]
@ceph/doc-writers can someone take a look at the docs

**Comment by anthonyeleven** [comment]
I don't see much in there for doc input

**Comment by cbodley** [comment]
passed qa in https://pulpito.ceph.com/cbodley-2024-03-19_01:03:50-rgw-wip-cbodley-testing-distro-default-smithi/ with rerun https://pulpito.ceph.com/cbodley-2024-03-19_12:32:32-rgw-wip-cbodley-testing-distro-default-smithi/

**Pull request merged**